src/tyr/serializers.cc

```diff
@@ -209,6 +237,47 @@ json::ArrayPtr waypoints(const valhalla::Trip& trip) {
   return waypoints;
 }
 
+/*
+ * This function takes any waypoints (excluding origin and destination) and walks the shape
+ * for a match.  When a match is found, we store the geometry index.
+ * We use the valhalla midgard `equal` function to check for equality based on precision.
+ */
+json::ArrayPtr via_waypoints(valhalla::Options options, const std::vector<PointLL>& shape) {
+  // Create a vector of indexes based on the number of locations.
+  auto locs = *options.mutable_locations();
+  uint32_t i = 0;
+  std::vector<uint32_t> indexes;
+  for (const auto& loc : locs) {
+    indexes.push_back(i++);
+  }
+  auto via_waypoints = json::array({});
+  for (const auto& index : indexes) {
+    // Only create via_waypoints object if the locations are via or through types
+    if (locs.Get(index).type() == 1 || locs.Get(index).type() == 2) {
+      locs.Mutable(index)->set_shape_index(index);
+      int geometry_idx = 0;
+
+      for (const auto& ll : shape) {
+        geometry_idx++;
+
+        if (index == 0 || index == (locs.size() - 1)) {
+          continue;
+        }
+        if (valhalla::midgard::equal<double>(static_cast<double>(
+                                                 locs.Get(index).path_edges(0).ll().lng()),
+                                             static_cast<double>(ll.first), .001) &&
+            valhalla::midgard::equal<double>(static_cast<double>(
+                                                 locs.Get(index).path_edges(0).ll().lat()),
+                                             static_cast<double>(ll.second), .001)) {
```

Comment:
A stylistic nit. This conditional is quite hard for me to read. What do you think about breaking it down into two `bool`s that are constructed before the if-statement?

Reply:
refactored